user-operator/src/main/java/io/strimzi/operator/user/operator/KafkaUserOperator.java

```diff
@@ -139,12 +151,125 @@ protected void createOrUpdate(Reconciliation reconciliation, KafkaUser kafkaUser
             scramAcls = user.getSimpleAclRules();
         }
 
+        KafkaUserStatus userStatus = new KafkaUserStatus();
         CompositeFuture.join(
                 scramShaCredentialOperator.reconcile(user.getName(), password),
-                secretOperations.reconcile(namespace, user.getSecretName(), desired),
+                reconcileSecretAndSetStatus(namespace, user, desired, userStatus),
                 aclOperations.reconcile(KafkaUserModel.getTlsUserName(userName), tlsAcls),
                 aclOperations.reconcile(KafkaUserModel.getScramUserName(userName), scramAcls))
-                .map((Void) null).setHandler(handler);
+                .compose(reconciliationResult -> {
```

Comment:
I didn't see a test for the `NotReady` case, and I think the use of `compose()` is incorrect. The lambda passed to `compose()` is only called then the receiver of the compose method complete normally. It's not called when the receiver completes with a failure. I think you need to use `setHandler()` instead of `compose()`, and then complete the given `handler` at the end of you lambda argument `setHandler()`.